### PR TITLE
refactor: compose complicated string with renderString

### PR DESCRIPTION
**Pull request by UncleBill**

### CODE REVIEW

**Comment by UncleBill** [comment]
Should be merged after #5871

**Comment by UncleBill** [comment]
Might rewrite with i18n instead.

File: packages/mask/src/plugins/Trader/SNSAdaptor/trader/Trader.tsx [hunk 1]
```diff
@@ -286,25 +287,31 @@ export function Trader(props: TraderProps) {
...
+    const cashTag = isOnTwitter ? '$' : ''
+    const maskAccount = isOnTwitter ? t('twitter_account') : t('facebook_account')
+    const promoteMask =
+        isOnTwitter || isFacebook(activatedSocialNetworkUI)
+            ? renderString(
+                  `Follow @{account} (mask.io) to swap cryptocurrencies on ${isOnTwitter ? 'Twitter' : 'Facebook'}`,
+                  {
+                      account: maskAccount,
+                  },
+              )
```
Comment (septs) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trader/Trader.tsx]:
```suggestion
            ? renderString(
                  `Follow @{account} (mask.io) to swap cryptocurrencies on {platform}`,
                  {
                      account: maskAccount,
                      platform: isOnTwitter ? 'Twitter' : 'Facebook'
                  },
              )
```

File: packages/shared-base/src/utils/renderString.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+type ExtraKeys<S extends string> = S extends `${infer HEAD}{${infer K}}${infer TAIL}`
+    ? K extends '}'
+        ? never
+        : ExtraKeys<HEAD> | K | ExtraKeys<TAIL>
+    : never
+
+export function renderString<T extends string>(template: T, data: Record<ExtraKeys<T>, string | number>) {
```
Comment (septs) [inline packages/shared-base/src/utils/renderString.ts:7]: suggest, rename the function name to `renderTemplateString`

Comment (guanbinrui) [reply packages/shared-base/src/utils/renderString.ts:7]: What about this already existing solution [_.template](https://lodash.com/docs/4.17.15#template)?

Comment (septs) [reply packages/shared-base/src/utils/renderString.ts:7]: lodash template function use eval function

Comment (UncleBill) [reply packages/shared-base/src/utils/renderString.ts:7]: We might rewrite with i18n instead. I18n engine itself provides template rendering. And these places miss i18n indeed.

**Comment by UncleBill** [comment]
Use i18n instead. Closing

**Comment by guanbinrui** [comment]
Close by https://github.com/DimensionDev/Maskbook/pull/5872#discussion_r833844461.

**Pull request closed without merging**